### PR TITLE
Implement _Alignas and _Alignof support with tests

### PR DESCRIPTION
**Pull request by vit9696**
Rebased the original PR on master.

### CODE REVIEW

File: pycparser/c_lexer.py [hunk 1]
```diff
@@ -111,7 +111,8 @@ def _make_tok_location(self, token):
 
     keywords_new = (
         '_BOOL', '_COMPLEX',
-        '_NORETURN', '_THREAD_LOCAL', '_STATIC_ASSERT', '_ATOMIC',
+        '_NORETURN', '_THREAD_LOCAL', '_STATIC_ASSERT',
+        '_ATOMIC', '_ALIGNOF', '_ALIGNAS',
```
Comment (eliben) [inline pycparser/c_lexer.py:115]: I'm not sure I see parser tests for `_Alignof` here. Did you have any?

File: pycparser/_c_ast.cfg [hunk 2]
```diff
@@ -172,14 +174,14 @@ TernaryOp: [cond*, iftrue*, iffalse*]
...
 # A base type declaration
 #
-TypeDecl: [declname, quals, type*]
+TypeDecl: [declname, quals, align, type*]
 
 # A typedef declaration.
 # Very similar to Decl, but without some attributes
 #
-Typedef: [name, quals, storage, type*]
+Typedef: [name, quals, align, storage, type*]
```
Comment (eliben) [inline pycparser/_c_ast.cfg:182]: Can typedefs have _Alignas?

File: pycparser/_ast_gen.py [hunk 3]
```diff
@@ -134,6 +134,22 @@ def _gen_children(self):
 
         return src
 
+    def _gen_eq(self):
```
Comment (eliben) [inline pycparser/_ast_gen.py:137]: Is this related to the PR?

File: pycparser/_ast_gen.py [hunk 4]
```diff
@@ -134,6 +134,22 @@ def _gen_children(self):
...
+        src  = '    def __eq__(self, o):\n'
+        src += '        if not o: return False\n'
```
Comment (eliben) [inline pycparser/_ast_gen.py:139]: Is this comparing children?

File: pycparser/_c_ast.cfg (hunk 2 above)
Comment (vit9696) [reply pycparser/_c_ast.cfg:182]: No.

File: pycparser/_ast_gen.py (hunk 3 above)
Comment (vit9696) [reply pycparser/_ast_gen.py:137]: Yes, required by tests.

File: pycparser/_ast_gen.py (hunk 4 above)
Comment (vit9696) [reply pycparser/_ast_gen.py:139]: Which particular case do you mean?

Comment (vit9696) [reply pycparser/_ast_gen.py:139]: `if self.{0} != o.{0}` will trigger recursive `__ne__` → `__eq__` calls. If you mean inner members.

File: pycparser/_ast_gen.py (hunk 3 above)
Comment (eliben) [reply pycparser/_ast_gen.py:137]: Could you please clarify which tests need this? pycparser has hundreds of AST tests which don't need this comparison, so I'm wondering what's new here. In any case I would prefer to decouple this large change from the `_Align*` PR.

File: tests/test_c_parser.py [hunk 5]
```diff
@@ -618,6 +623,12 @@ def test_sizeof(self):
                         ['TypeDecl',
                             ['IdentifierType', ['int']]]]]])
 
+    def test_alignof(self):
```
Comment (eliben) [inline tests/test_c_parser.py:626]: I'd like to see more parser tests here.

At least one test separately for `_Alignof` without `_Alignas` (look at the existing `sizeof` tests)
A test for `_Alignas` with a true constant like `4`
A test for `_Alignas` with a type inside e.g. `_Alignas(int)`.

File: pycparser/c_parser.py [hunk 6]
```diff
@@ -310,6 +310,11 @@ def _fix_decl_name_type(self, decl, typename):
         decl.name = type.declname
         type.quals = decl.quals[:]
 
+        for tn in typename:
+            if isinstance(tn, c_ast.Alignas):
+                type.align = tn
```
Comment (eliben) [inline pycparser/c_parser.py:315]: I'm a bit confused here: isn't align an attribute, not a child node? Should it be a child node instead if you're assigning nodes to it?

File: pycparser/c_parser.py [hunk 7]
```diff
@@ -310,6 +310,11 @@ def _fix_decl_name_type(self, decl, typename):
...
+                typename.remove(tn)
```
Comment (eliben) [inline pycparser/c_parser.py:316]: Don't modify a list you're iterating.

Comment (eliben) [reply pycparser/c_parser.py:316]: Also, if we allow a list of alignment specifiers, does it show here or do we only overwrite them?

File: pycparser/c_parser.py [hunk 8]
```diff
@@ -920,9 +940,20 @@ def p_specifier_qualifier_list_3(self, p):
     def p_specifier_qualifier_list_4(self, p):
         """ specifier_qualifier_list  : type_qualifier_list type_specifier
         """
-        spec = dict(qual=p[1], storage=[], type=[], function=[])
+        spec = dict(qual=p[1], alignment=[], storage=[], type=[], function=[])
```
Comment (eliben) [inline pycparser/c_parser.py:943]: So you really need _add_declaration_specifier here? Can't you just copy the `type`?

File: pycparser/c_parser.py [hunk 9]
```diff
@@ -920,9 +940,20 @@ def p_specifier_qualifier_list_3(self, p):
...
         p[0] = self._add_declaration_specifier(spec, p[2], 'type', append=True)
 
+    def p_specifier_qualifier_list_5(self, p):
+        """ specifier_qualifier_list  : alignment_specifier
+        """
+        spec = dict(qual=[], alignment=[], storage=[], type=[], function=[])
+        p[0] = self._add_declaration_specifier(spec, p[1], 'alignment')
```
Comment (eliben) [inline pycparser/c_parser.py:950]: Same question here, it's not clear why you need to create an empty `spec` and pass it to _add_declaration_specifier

**Review by eliben — CHANGES_REQUESTED** [review]

File: pycparser/_ast_gen.py (hunk 3 above)
Comment (vit9696) [reply pycparser/_ast_gen.py:137]: This one. I am fine to create a separate PR for that.

```
======================================================================
FAIL: test_alignment (tests.test_c_generator.TestCtoC)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/user/AppleDeveloper/pycparser/tests/test_c_generator.py", line 100, in test_alignment
    self._assert_ctoc_correct('_Alignas(32) int b;')
  File "/Users/user/AppleDeveloper/pycparser/tests/test_c_generator.py", line 79, in _assert_ctoc_correct
    "{!r} != {!r}".format(src, src2))
AssertionError: '_Alignas(32) int b;' != '_Alignas(32) int b;\n'
```

File: tests/test_c_parser.py (hunk 5 above)
Comment (vit9696) [reply tests/test_c_parser.py:628]: Done, thanks.

File: pycparser/c_parser.py (hunk 6 above)
Comment (vit9696) [reply pycparser/c_parser.py:315]: It is an attribute indeed, not a child node. This code was merely a fixup for the parser limitation. Previously one of the tests triggered an issue without this code, however, currently none does. I am in a good faith to believe that the issue was resolved more properly in one of the future iterations of the code and I did not notice it. Will simply remove.

File: pycparser/c_parser.py (hunk 7 above)
Comment (vit9696) [reply pycparser/c_parser.py]: Fixed.

File: pycparser/c_parser.py (hunk 8 above)
Comment (vit9696) [reply pycparser/c_parser.py]: Rebased on top of master fixing. Indeed no need.

File: pycparser/c_parser.py (hunk 9 above)
Comment (vit9696) [reply pycparser/c_parser.py]: Sounds like literal copy-pasting from the existing code, nice catch. Also fixed.

File: pycparser/_ast_gen.py (hunk 3 above)
Comment (eliben) [reply pycparser/_ast_gen.py:137]: Alright, let's move the equality stuff out to a separate PR that can be discussed after this one lands.

**Review by eliben — CHANGES_REQUESTED** [review]
Thanks, this looks close just. Just remove the equality generation for now, and I'll take a final look

**Comment by vit9696** [comment]
Sure, shall I comment the test part for now?

**Comment by eliben** [comment]
> Sure, shall I comment the test part for now?

No, please remove everything related to generating equality and the test that requires it from the PR. I'm still not convinced this is needed at all, and wanted to separate it to its own PR so as not to block the align feature

**Comment by vit9696** [comment]
> No, please remove everything related to generating equality and the test that requires it from the PR. I'm still not convinced this is needed at all, and wanted to separate it to its own PR so as not to block the align feature

Hmm, this is the whole `def test_alignment` function though. There is no single test step that passes without the `__eq__` functions. Updated the PR.

File: tests/test_c_generator.py [hunk 10]
```diff
@@ -96,6 +96,27 @@ def test_complex_decls(self):
         self._assert_ctoc_correct('int test(const char* const* arg);')
         self._assert_ctoc_correct('int test(const char** const arg);')
 
+    # FIXME: These require custom equality comparison.
```
Comment (vit9696) [inline tests/test_c_generator.py:99]: I can remove that block if you think it is better.

**Review by eliben — APPROVED** [review]

**Comment by eliben** [comment]
I've re-enabled the commented-out tests in d4c2922407c486da1ba9c6cd769990cc64308be6

**Comment by eliben** [comment]
Thank you for pushing the align and other changes through and improving C11 support for pycparser :)
